File: python/MedianSortedArrays.py
from typing import List

class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        A, B = nums1, nums2
        if len(B) < len(A):
            A, B = B, A

        total = len(A) + len(B)
        half = total // 2

        l, r = 0, len(A) - 1

        while True:
            i = l + r // 2
            j = half - i - 2

            aleft = A[i] if i >= 0 else float("-infinity")
            aright = A[i + 1] if i + 1 < len(A) else float("infinity")
            bleft = B[j] if j >= 0 else float("-infinity")
            bright = B[j + 1] if j + 1 < len(B) else float("infinity")

            if aleft <= bright and bleft <= aright:
                if total % 2:
                    return min(aright, bright)
                else:
                    return (max(aleft, bleft) + min(aright, bright)) / 2.0
            elif aleft > bright:
                r = i - 1
            else:
                r = i + 1

if __name__ == '__main__':
    nums1 = [1, 2]
    nums2 = [3, 4]
    print(Solution().findMedianSortedArrays(nums1, nums2))

File: src/com/soulmeadows/MedianSortedArrays.java
package com.soulmeadows;

public class MedianSortedArrays {

    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        if (nums1.length > nums2.length) {
            return findMedianSortedArrays(nums2, nums1);
        }

        int m = nums1.length;
        int n = nums2.length;
        int total = m + n;
        int half = total / 2;

        int l = 0;
        int r = m;
        while(true){
            int i = (l + r) / 2; // A
            int j = half  - i - 2; // B

            int aleft = (i >= 0) ? nums1[i]: Integer.MIN_VALUE;
            int aright = (i + 1 < m) ? nums1[i + 1]: Integer.MAX_VALUE;
            int bleft = (j >= 0) ?  nums2[j] : Integer.MIN_VALUE;
            int bright = (j +1 < n) ? nums2[j + 1]: Integer.MAX_VALUE;

            if(aleft <= bright && bleft <= aright){
                if (total % 2 == 1){
                    return Math.max(aleft, bleft);
                }else{
                    return (Math.max(aleft, bleft) + Math.min(aright, bright)) / 2.0;
                }
            }else if(aleft > bright){
                r = i - 1;
            }else {
                l = i + 1;
            }
        }
        //return 0.0;

    }

    public static void main(String[] args){
        int[] nums1 = {1, 3};
        int[] nums2 = {2};
        new MedianSortedArrays().findMedianSortedArrays(nums1, nums2);
    }
}
